src/Server/WebSocketServer.php

```
@@ -162,10 +164,51 @@ private function onNewConnection(ConnectionInterface $socketStream)
             return $this->getMessageHandler($uri, $connection);
         }, $this->loop, $this->messageProcessor);
 
+        $socketStream->on('end', function () use($connection) {
+            $this->onDisconnect($connection);
+        });
+
         $connection->setLogger($this->getLogger());
+        $connection->getLogger()->info(sprintf('Ip "%s" establish connection at "%s" UTC',
+            $connection->getIp(),
+            date(self::FULL_DATE_FORMAT)
+        ));
         $this->connections[] = $connection;
     }
 
+    /**
+     *
+     * @param Connection $connection
+     */
+    private function onDisconnect(Connection $connection)
+    {
+        $connection->disconnect();
+        $this->removeConnection($connection);
+        $connection->getLogger()->info(sprintf('Ip "%s" left connection at "%s" UTC',
+            $connection->getIp(),
+            date(self::FULL_DATE_FORMAT)
+        ));
+
+        unset($connection);
+    }
+
+    /**
+     * Remove a Connection instance by his object id, or log not found
+     * @param Connection $connection
+     */
+    private function removeConnection(Connection $connection)
+    {
+        $connectionId = spl_object_hash($connection);
+        foreach ($this->connections as $index => $connectionIt) {
+            if ($connectionId === spl_object_hash($connectionIt)) {
+                unset($this->connections[$index]);
+                return;
+            }
+        }
+
+        $this->logger->warning(sprintf('Impossible to find the connection with id "%d" in the running server.', $connectionId));
```

You should throw a runtime exception at this point. Nothing is normal here.